Add support for TFA Dostmann 14.1504 Radio-controlled grill and meat thermometer (#2288)

joel-bourquard opened the pull request.

zuckschwerdt commented on src/devices/tfa_14_1504.c:
```diff
@@ -0,0 +1,199 @@
+*/
+
+/**
+Decoder for temperature sensor model no. TPX305
```
This should include (replicate) line 2 infos.

joel-bourquard replied:
Done! Thanks

zuckschwerdt commented:
That checksum looks accidental. A wide enough CRC can always be produced for just a few bits.
Are you sure it's a CRC? Maybe it's a digest (LFSR)? Can you share the list of captured codes you derived this from?

joel-bourquard commented:
> That checksum looks accidental. A wide enough CRC can always be produced for just a few bits. Are you sure it's a CRC? Maybe it's a digest (LFSR)? Can you share the list of captured codes you derived this from?

I know, I can tell you that I tried a few exhaustive searches with the 16-bit LFSR functions from util.c (rtl_433), more precisely the lfsr_digest16() function with all possible combinations for "gen" and "key", and applied that to all possible substrings... with no success there. I was very surprised/frustrated too, but maybe I missed something obvious.

Sure, here are all samples along with the tool to compare the computed CRCs against them. You can simply clone this and run "./build.sh":
https://github.com/joel-bourquard/tfa_tpx305_reveng

zuckschwerdt commented:
Did you try our [revdgst16](https://github.com/triq-org/revdgst/)? It finds gen 8810 key 93ad final xor d17c when you omit the preamble.
0x8810 is *the* gen used everywhere, which sounds reassuring, the final xor could mean that we picket up some sync that shouldn't be used -- not sure.

joel-bourquard commented:
> Did you try our [revdgst16](https://github.com/triq-org/revdgst/)? It finds gen 8810 key 93ad final xor d17c when you omit the preamble. 0x8810 is _the_ gen used everywhere, which sounds reassuring, the final xor could mean that we picket up some sync that shouldn't be used -- not sure.

Aah awesome, thanks a lot! No I didn't know that tool, works great indeed. The reason my prior brute force attempts didn't work out, was that final XOR.

I've tried a few things but no zero final XOR so far - - so I've updated my PR with the LSFR and the values that you mentioned - which I think are the most suitable so far.

zuckschwerdt commented on src/devices/tfa_14_1504.c:
```diff
@@ -0,0 +1,132 @@
+        return DECODE_FAIL_MIC;
+    }
+
+    uint16_t raw_temp_c = ((data[1] & 0xf) << 6) + (data[2] >> 2);
```
Are you shifting off the last 2 bits to avoid a div 4? Are those never used, i.e. is the precission not 1/4 deg C?

joel-bourquard replied:
> Are you shifting off the last 2 bits to avoid a div 4? Are those never used, i.e. is the precission not 1/4 deg C?

Yes exactly: for each degree C, the value increases by 4. The two lower bits are always zero, as far as I've seen (and it's also seen in the samples.txt file). Also the display on the device itself does not have a fractional part.
For context, it's a barbecue thermometer with a high temperature probe which goes up to 300 degrees C, but is probably not precise enough to justify having a fractional part...

zuckschwerdt replied:
Ok, good. Add a `// stripping off last 2 bit as those are always zero` or such ;)  But a float div 4 shoould not be expensive either, as it likely only subs 2 from the exponent, no mantissa manipulation needed.

joel-bourquard replied:
Done!

zuckschwerdt commented:
If possible I wouldn't use the full preamble but just a tail portion. This will improve reception if the signal isn't immediatly demodulated perfectly, how typical receivers work to "tune-in". That scheme needs a "sync" at the end to known when data starts, could be a single bit, but is usually aligned. So if `5c` is hard coded to be matched anyway then you can use a robust preamble of e.g. `aaaa5c`.

joel-bourquard commented:
> If possible I wouldn't use the full preamble but just a tail portion. This will improve reception if the signal isn't immediatly demodulated perfectly, how typical receivers work to "tune-in". That scheme needs a "sync" at the end to known when data starts, could be a single bit, but is usually aligned. So if `5c` is hard coded to be matched anyway then you can use a robust preamble of e.g. `aaaa5c`.

Good point thanks, I will do that!

Question about the Probe status: currently I set it to "Connected" / "Disconnected", is this good? Or should I use a different upper/lowercase convention, or an int (0/1) for that?

zuckschwerdt commented:
> Question about the Probe status: currently I set it to "Connected" / "Disconnected", is this good? Or should I use a different upper/lowercase convention, or an int (0/1) for that?

I would use a 0/1 int for easier transfer into databases (int supports aggregation), like the common `battery_ok`, so `probe_ok`. Or perhaps even invert the output and name it `probe_fail`? We don't have a common scheme for these "event like" data keys.

joel-bourquard commented:
> > Question about the Probe status: currently I set it to "Connected" / "Disconnected", is this good? Or should I use a different upper/lowercase convention, or an int (0/1) for that?
> 
> I would use a 0/1 int for easier transfer into databases (int supports aggregation), like the common `battery_ok`, so `probe_ok`. Or perhaps even invert the output and name it `probe_fail`? We don't have a common scheme for these "event like" data keys.

Good point! I think probe_fail is the best option, and pushed the change now!
(user-friendly text is "Probe failed").

joel-bourquard commented:
Battery flag implemented and works well. No more changes planned, so if you can consider it for inclusion, it would be great 😉
If any other changes are advised or if I should squash, please let me know...

joel-bourquard commented:
Hi @zuckschwerdt,
I've been run-testing the decoder for a few days now, under challenging SNR conditions, and it's working great so far. No other changes planned for me at this point.

What do you think about merging my PR?

I'm also looking forward to getting feedback from other owners of this device, to validate whether all LSFR params are identical or not. If they are not, then I have some ideas on how to address it.

zuckschwerdt commented:
Yes, looks good. The "mic" should be "CRC" as it's the same strength.
And "model" should be simpler, e.g. "TFA-141504v2" (only one dash in the middle, no other special chars).

joel-bourquard commented:
> Yes, looks good. The "mic" should be "CRC" as it's the same strength. And "model" should be simpler, e.g. "TFA-141504v2" (only one dash in the middle, no other special chars).

Great, thanks! I've implemented the exact changes that you suggested and squashed my commits. Ready to merge for me 😉

joel-bourquard commented:
Hi @zuckschwerdt ,
Great, thanks! I've implemented the exact changes that you suggested and squashed my commits. Ready to merge for me 😉

joel-bourquard commented:
Hi @zuckschwerdt,
Do you know when my PR can be merged? Or should I ping someone?
Thanks 😉

merbanan commented on src/devices/tfa_14_1504_v2.c:
```diff
@@ -0,0 +1,141 @@
+    uint8_t const preamble[] = {
+            0xaa,
+            0xaa,
+            0x5c,
```
Less lines if all bytes are in a row.

joel-bourquard replied:
Hi @merbanan , thank you!   Change applied: https://github.com/merbanan/rtl_433/pull/2296/commits/d145440875ebbc91f26b5ceb329072f81a18f51a

merbanan commented on src/devices/tfa_14_1504_v2.c:
```diff
@@ -0,0 +1,141 @@
+    // we discard the last 2 bits as those are always zero
+    uint16_t raw_temp_c = ((data[0] & 0xf) << 6) + (data[1] >> 2);
+    unsigned is_probe_connected = (raw_temp_c != 0x1c0);
+    double temp_c = ((int)raw_temp_c) - 532;
```
Should be float, besides that

joel-bourquard replied:
Change applied: https://github.com/merbanan/rtl_433/pull/2296/commits/d145440875ebbc91f26b5ceb329072f81a18f51a

merbanan commented:
LGTM, but I think we we need docs and config changes also. And then some sample signals.

merbanan commented:
@joel-bourquard 
Did you upload any signal samples to the test repo ? I can not see any pull request.

https://github.com/merbanan/rtl_433_tests/pulls

merbanan commented:
In this commit you can see the needed config and documentation changes.

https://github.com/merbanan/rtl_433/commit/cd90897b7f7359423058f392c9e786d96ca4c3b8

joel-bourquard commented:
> In this commit you can see the needed config and documentation changes.
> 
> [cd90897](https://github.com/merbanan/rtl_433/commit/cd90897b7f7359423058f392c9e786d96ca4c3b8)

Should be all done, thanks!  => https://github.com/merbanan/rtl_433/pull/2296/commits/d145440875ebbc91f26b5ceb329072f81a18f51a

joel-bourquard commented:
> @joel-bourquard Did you upload any signal samples to the test repo ? I can not see any pull request.
> 
> https://github.com/merbanan/rtl_433_tests/pulls

@merbanan not yet, but I will definitely do it!  I'll fork that repo and do a PR.

joel-bourquard commented:
Hi @zuckschwerdt , @merbanan ,
All ready on the rtl_433 side, I think :-)

So I have created a PR with documentation+samples on the rtl_433_tests side:
https://github.com/merbanan/rtl_433_tests/pull/445

joel-bourquard commented:
Rebased on latest master.

joel-bourquard commented:
Hi @zuckschwerdt ,
Rebased on latest master.  Would it be possible for you to merge it please? :)

zuckschwerdt commented:
Just a few other (~3) merges first.
